### PR TITLE
Health page

### PR DESCRIPTION
**Pull request by Boy132**
Uses [spatie/laravel-health](https://github.com/spatie/laravel-health) to add a health page.
Also includes two custom checks: one for checking the panel version and one for checking node versions.

grafik
grafik

### CODE REVIEW

**Review by RMartinOscar — COMMENTED** [review]
Would be nice if this could be set to the user's prefered theme so we don't get *Throwing flash bang*
> Also applies to Scramble's docs

Other then that its dope very concise

File: app/Checks/NodeVersionsCheck.php [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+        $outdated = 0;
+
+        $latestVersion = app(SoftwareVersionService::class)->getDaemon();
+
+        $nodes = Node::query()->get();
+        foreach ($nodes as $node) {
+            if ($node->systemInformation()['version'] !== $latestVersion) {
+                $outdated++;
+            }
+        }
```
Comment (RMartinOscar) [inline app/Checks/NodeVersionsCheck.php:24]:
```suggestion
        $outdated = Node::query()->get()
            ->filter(function (Node $node, bool $latestVersion) {
                $systemInformation = $node->systemInformation();
                return !isset($systemInformation['exception']) && $systemInformation['version'] !== $latestVersion;
            })
            ->count();
```

Comment (RMartinOscar) [inline app/Checks/NodeVersionsCheck.php:24]:
```suggestion
        $latestVersion = app(SoftwareVersionService::class)->getDaemon();

         $outdated = Node::query()->get()
            ->filter(fn (Node $node) => !isset($node->systemInformation()['exception']) && $node->systemInformation()['version'] !== $latestVersion)
            ->count();
```

**Review by RMartinOscar — COMMENTED** [review]
Just tried it the page is cleared until i hit refresh atleast once or wait [5min](https://github.com/pelican-dev/panel/pull/469/files#diff-3e91b9df26f6ed49ead694c97ddcddf5f3c3e878390064d0b8849f22d54e5603R47) since build.
I would remove that unless anyway cause i want to know if debug is on even on local env just an other safe check doesn't hurt imo

**Comment by notAreYouScared** [comment]
Since disk space does not work on windows do we have a way to detect that? Many not show it as crashed?

File: app/Providers/AppServiceProvider.php [hunk 2]
```diff
@@ -80,6 +89,20 @@ public function boot(): void
         Event::listen(function (\SocialiteProviders\Manager\SocialiteWasCalled $event) {
             $event->extendSocialite('discord', \SocialiteProviders\Discord\Provider::class);
         });
+
+        // Don't run any health checks during tests
+        if (config('app.env') !== 'testing') {
```
Comment (lancepioch) [inline app/Providers/AppServiceProvider.php:94]:
```suggestion
        if (!app()->runningUnitTests()) {
```

File: app/Providers/AppServiceProvider.php [hunk 3]
```diff
@@ -80,6 +89,20 @@ public function boot(): void
...
+            Health::checks([
+                DebugModeCheck::new()->unless(app()->environment('local')),
```
Comment (lancepioch) [inline app/Providers/AppServiceProvider.php:96]: unless !app()->isProduction()
or
if app()->isProduction

**Review by lancepioch — CHANGES_REQUESTED** [review]
Looks good, got a couple nitpicks

File: app/Providers/AppServiceProvider.php (hunk 3 above)
Comment (Boy132) [reply app/Providers/AppServiceProvider.php:96]:
```suggestion
                DebugModeCheck::new()->unless(!app()->isProduction()),
```

**Comment by Boy132** [comment]
> Since disk space does not work on windows do we have a way to detect that? Many not show it as crashed?

We don't support windows anyways?

**Comment by notAreYouScared** [comment]
> > Since disk space does not work on windows do we have a way to detect that? Many not show it as crashed?
> 
> We don't support windows anyways?

Since when can the panel itself not be ran on Windows? I have my dev env setup using herd... on Windows

**Comment by Boy132** [comment]
> > > Since disk space does not work on windows do we have a way to detect that? Many not show it as crashed?
> > 
> > 
> > We don't support windows anyways?
> 
> Since when can the panel itself not be ran on Windows? I have my dev env setup using herd... on Windows

Unsupported ≠ not possible. Officially there is no Windows support, it's not listed as supported os in the docs.

**Comment by RMartinOscar** [comment]
I think [this PR](https://github.com/spatie/laravel-health/pull/132/files) would fix the windows issue but didn't get merged.
> Feel free to create a custom health check in your own project using the native functions.

File: app/Providers/AppServiceProvider.php [hunk 4]
```diff
@@ -2,6 +2,8 @@
 
 namespace App\Providers;
 
+use App\Checks\NodeVersionsCheck;
```
Comment (RMartinOscar) [inline app/Providers/AppServiceProvider.php:5]:
```suggestion
use App\Checks\NodeVersionsCheck;
use App\Checks\DiskUsageCheck;
```

File: app/Providers/AppServiceProvider.php [hunk 5]
```diff
@@ -21,6 +23,13 @@
 use Illuminate\Support\ServiceProvider;
 use Illuminate\Support\Str;
 use Laravel\Sanctum\Sanctum;
+use Spatie\Health\Checks\Checks\CacheCheck;
+use Spatie\Health\Checks\Checks\DatabaseCheck;
+use Spatie\Health\Checks\Checks\DebugModeCheck;
+use Spatie\Health\Checks\Checks\EnvironmentCheck;
+use Spatie\Health\Checks\Checks\ScheduleCheck;
+use Spatie\Health\Checks\Checks\UsedDiskSpaceCheck;
```
Comment (RMartinOscar) [inline app/Providers/AppServiceProvider.php:31]:
```suggestion
```

File: app/Providers/AppServiceProvider.php [hunk 6]
```diff
@@ -80,6 +89,20 @@ public function boot(): void
...
+
+        // Don't run any health checks during tests
+        if (!app()->runningUnitTests()) {
+            Health::checks([
+                DebugModeCheck::new()->if(app()->isProduction()),
+                EnvironmentCheck::new(),
+                CacheCheck::new(),
+                DatabaseCheck::new(),
+                ScheduleCheck::new(),
+                UsedDiskSpaceCheck::new(),
```
Comment (RMartinOscar) [inline app/Providers/AppServiceProvider.php:101]:
```suggestion
                DiskUsageCheck::new(),
```

**Review by RMartinOscar — COMMENTED** [review]
Following https://github.com/pelican-dev/panel/pull/469#issuecomment-2224980544

So maybe create a `app/Checks/DiskUsageCheck.php`
```suggestion

<?php

namespace App\Checks;

use Spatie\Health\Checks\Checks\UsedDiskSpaceCheck;

class DiskUsageCheck extends UsedDiskSpaceCheck
{
    protected function getDiskUsagePercentage(): int
    {
        $freeSpace = disk_free_space($this->filesystemName ?? '/');
        $totalSpace = disk_total_space($this->filesystemName ?? '/');

        return (int) 100 - ($freeSpace * 100 / $totalSpace);
    }
}```

**Comment by notAreYouScared** [comment]
Hey @Boy132, while reviewing your PR, I'd suggest the following code changes:

👉 [**Code Suggestion for #469**](https://gitkraken.dev/drafts/f8b283d3-81ed-4350-af65-e751d29d2c8e?source=PRComment)
>https://github.com/pelican-dev/panel/pull/469

You can also review and apply these suggestions [locally on your machine](https://gitkraken.dev/link/drafts/f8b283d3-81ed-4350-af65-e751d29d2c8e?type=suggested_pr_change&prEntityId=WyJnaXRodWIiLCJwciIsIjEiLCIiLCJQUl9rd0RPTGlaQkU4NTBzSjZvIl0=&source=PRComment).

<details><summary>Learn more about GitKraken Code Suggest</summary>

>[Code Suggest](https://gitkraken.dev/redirect/aHR0cHM6Ly93d3cuZ2l0a3Jha2VuLmNvbS9zb2x1dGlvbnMvY29kZS1zdWdnZXN0P3NvdXJjZT1naXRodWIuY29tJm1lZGl1bT1wcl9jb21tZW50?source=PRSuggest&event=redirectToCodeSuggestPage&provider=github) liberates your code reviews from GitHub's restrictive, comment-only feedback style. As simple as suggesting changes in a Google-doc, provide real code suggestions from where you code, e.g. your IDE, and on anything in your project — not just on the lines of code changed in the PR.

_Join your team on [GitKraken](https://gitkraken.dev/redirect/aHR0cHM6Ly93d3cuZ2l0a3Jha2VuLmNvbT9zb3VyY2U9Z2l0aHViLmNvbSZtZWRpdW09cHJfY29tbWVudA==?source=PRSuggest&event=redirectToIndexPage&provider=github) to speed up PR review._
</details>

File: app/Filament/Pages/HealthCheckResults.php [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+            $this->dispatch('refresh-component');
+        }
+
+        return [
+            'lastRanAt' => new Carbon($checkResults?->finishedAt),
+            'checkResults' => $checkResults,
+        ];
+    }
+}
```
Comment (notAreYouScared) [inline app/Filament/Pages/HealthCheckResults.php:35]:
```suggestion
<?php

namespace App\Filament\Pages;

use Carbon\Carbon;
use Illuminate\Support\Facades\Artisan;
use ShuvroRoy\FilamentSpatieLaravelHealth\Pages\HealthCheckResults as BaseHealthCheckResults;
use Spatie\Health\Commands\RunHealthChecksCommand;
use Spatie\Health\ResultStores\ResultStore;

class HealthCheckResults extends BaseHealthCheckResults
{
    protected static ?string $slug = 'health';
    protected static ?int $navigationSort = 4;

    protected function getViewData(): array
    {
        $checkResults = app(ResultStore::class)->latestResults();

        if ($checkResults === null) {
            Artisan::call(RunHealthChecksCommand::class);

            $this->dispatch('refresh-component');
        }

        return [
            'lastRanAt' => new Carbon($checkResults?->finishedAt),
            'checkResults' => $checkResults,
        ];
    }

    public static function getNavigationGroup(): ?string
    {
        return 'Advanced';
    }

    public static function getNavigationLabel(): string
    {
        return 'Panel Health';
    }

    public static function getNavigationBadge(): ?string
    {
        $results = json_decode(app(ResultStore::class)->latestResults()->toJson(), true);

        $failed = array_reduce($results['checkResults'], function ($numFailed, $result) {
            return $numFailed + ($result['status'] === 'failed' ? 1 : 0);
        }, 0);

        return $failed === 0 ? null : (string) $failed;
    }

    public static function getNavigationBadgeColor(): string
    {
        return self::getNavigationBadge() > null ? 'danger' : '';
    }

    public static function getNavigationBadgeTooltip(): ?string
    {
        $results = json_decode(app(ResultStore::class)->latestResults()->toJson(), true);

        $failedNames = array_reduce($results['checkResults'], function ($carry, $result) {
            if ($result['status'] === 'failed') {
                $carry[] = $result['name'];
            }

            return $carry;
        }, []);

        return 'Failed: ' . implode(', ', $failedNames);
    }

    public static function getNavigationIcon(): string
    {
        return app(ResultStore::class)->latestResults()->containsFailingCheck() ? 'tabler-heart-exclamation' : 'tabler-heart-check';
    }

    public static function getSlug(): string
    {
        return 'health';
    }
}
```

**Review by notAreYouScared — APPROVED** [review]

**Comment by Boy132** [comment]
I want to remove [shuvroroy/filament-spatie-laravel-health](https://github.com/shuvroroy/filament-spatie-laravel-health) but for some reason tailwind classes aren't working with the custom page/ view:
grafik

File: app/Providers/AppServiceProvider.php [hunk 8]
```diff
@@ -115,6 +115,10 @@ public function boot(): void
                 NodeVersionsCheck::new(),
             ]);
         }
+
+        Gate::before(function (User $user, $ability) {
+            return $user->isRootAdmin() ? true : null;
+        });
```
Comment (RMartinOscar) [inline app/Providers/AppServiceProvider.php:121]:
```suggestion
        Gate::before(fn (User $user) => $user->isRootAdmin() ? true : null);
```
Removed unused var and used arrow function

Comment (Boy132) [reply app/Providers/AppServiceProvider.php:121]: That's a commit from main, not from this PR.

**Comment by notAreYouScared** [comment]
image

Any way to remove the `'s? As they don't render in code blocks and kind of look... un-needed?

File: app/Checks/NodeVersionsCheck.php [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+        $result = Result::make()
+            ->meta([
+                'all' => $all,
+                'outdated' => $outdated,
+            ])
+            ->shortSummary($outdated === 0 ? 'All up-to-date' : "{$outdated}/{$all} outdated");
+
+        return $outdated === 0
+            ? $result->ok('All Nodes are up-to-date')
+            : $result->failed("`{$outdated}`/`{$all}` Nodes are outdated");
```
Comment (notAreYouScared) [inline app/Checks/NodeVersionsCheck.php:40]:
```suggestion
            : $result->failed("{$outdated}/{$all} Nodes are outdated");
```

File: app/Checks/PanelVersionCheck.php [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+            ->meta([
+                'isLatest' => $isLatest,
+                'currentVersion' => $currentVersion,
+                'latestVersion' => $latestVersion,
+            ])
+            ->shortSummary($isLatest ? 'up-to-date' : 'outdated');
+
+        return $isLatest
+            ? $result->ok('Panel is up-to-date')
+            : $result->failed("Installed version is `{$currentVersion}` but latest is `{$latestVersion}`");
```
Comment (notAreYouScared) [inline app/Checks/PanelVersionCheck.php:30]:
```suggestion
            : $result->failed("Installed version is {$currentVersion} but latest is {$latestVersion}");
```

File: app/Checks/UsedDiskSpaceCheck.php [hunk 11]
```diff
@@ -0,0 +1,16 @@
+<?php
```
Comment (notAreYouScared) [inline app/Checks/UsedDiskSpaceCheck.php:1]: Should we check the users locale for number formatting and GiB/GB?

**Review by notAreYouScared — CHANGES_REQUESTED** [review]

**Comment by Boy132** [comment]
> image
> 
> Any way to remove the `'s? As they don't render in code blocks and kind of look... un-needed?
> 
> Other than that, looks good

For other custom checks yes, but I don't think there is a way to remove them in the built-in checks.

**Review by notAreYouScared — APPROVED** [review]

**Pull request merged**